pkg/cloud/gcp/actuators/machine/reconciler.go

```diff
@@ -66,8 +66,22 @@ func containsString(sli []string, str string) bool {
 	return false
 }
 
-func restartPolicyToBool(policy machinev1.GCPRestartPolicyType) *bool {
-	restart := policy == machinev1.RestartPolicyAlways
+func restartPolicyToBool(policy machinev1.GCPRestartPolicyType, preemptible bool) *bool {
+	// for more information about how the restart policy works, see the GCP docs at
+	// https://cloud.google.com/compute/docs/instances/setting-vm-host-options#settingoptions
+	restart := false
+
+	if preemptible {
+		// preemptible instances cannot be set to always restart
+		klog.V(4).Info("Setting preemptible instance restart policy to never")
+		return &restart
+	}
+
+	if policy == machinev1.RestartPolicyAlways || len(policy) == 0 {
+		// if the policy is unset (zero value), we use Always as the default as specified by the GCP docs
+		restart = true
```

Comment:
Do we need to inline GCP's upstream choice?  From [GCP's docs][1]:

> Compute Engine configures all VMs with default values unless you specify otherwise.

I'd expect that to be something like:

```go
if len(policy) == 0 {
  return nil, nil // caller had no opinion, so let the cloud provider pick a default
} else if policy == machinev1.RestartPolicyAlways {
  if preemptible {
    return nil, errors.New("preemptible instances cannot be automatically restarted") // caller can complain about this with a Machine reference, so the log-reading user can follow the trail of the impossible request.
  }
  return pointer.Bool(true) // caller requested automatic restarts
} else if policy == machinev1.RestartPolicyNever {
  return pointer.Bool(false) // caller requested manual restarts
}
return nil, fmt.Errorf("unrecognized restart policy %q", policy) // caller can complain about this with a Machine reference, so the log-reading user can follow the trail of the impossible request.
```

or something?

[1]: https://cloud.google.com/compute/docs/instances/host-maintenance-options#settingoptions

Reply:
hmm, yeah, perhaps i went too deep replicating the logic. i guess we can pass through the `nil` values when it is not specified and allow the platform to make the decision.